### ISSUE TITLE
undefined method `valid_encoding?' on upgrading to 1.1.0

### ISSUE THREAD

**Issue by holden** [wingrunr21/griddler-mandrill#18]
I upgraded to 1.1.0 to try and use headers (from 1.0.2) but before I got that far I am now getting:  (re: https://github.com/wingrunr21/griddler-mandrill/pull/12)

```
NoMethodError (undefined method `valid_encoding?' for #<Array:0x007feacd574bc0>
```

Upgraded to 1.1.0 for griddler-mandrill other than that nothing changed (griddler is 1.2.0 on both). It still works fine for me when I downgrade back to 1.0.2. Did anything else change?

I tried boiling it down to the most basic thing, but even this breaks:

```
class EmailProcessor
  def initialize(email)
    @email = email
  end

  def process
  	@email.to.each do |recipient|
  		@account = EmailAccount.find_by(email: recipient[:email])
  	end

    #more logic here.
   end
end
```

```
NoMethodError (undefined method `valid_encoding?' for #<Array:0x007fe651a95328>):
  griddler (1.2.0) lib/griddler/email.rb:83:in `clean_invalid_utf8_bytes'
  griddler (1.2.0) lib/griddler/email.rb:54:in `block in extract_headers'
  griddler (1.2.0) lib/griddler/email.rb:53:in `each'
  griddler (1.2.0) lib/griddler/email.rb:53:in `inject'
  griddler (1.2.0) lib/griddler/email.rb:53:in `extract_headers'
  griddler (1.2.0) lib/griddler/email.rb:21:in `initialize'
  griddler (1.2.0) app/controllers/griddler/emails_controller.rb:4:in `new'
  griddler (1.2.0) app/controllers/griddler/emails_controller.rb:4:in `block in create'
  griddler (1.2.0) app/controllers/griddler/emails_controller.rb:3:in `each'
  griddler (1.2.0) app/controllers/griddler/emails_controller.rb:3:in `create'
  actionpack (4.2.0) lib/action_controller/metal/implicit_render.rb:4:in `send_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:198:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rendering.rb:10:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:20:in `block in process_action'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_process_action_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/abstract_controller/callbacks.rb:19:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/rescue.rb:29:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `block in instrument'
  activesupport (4.2.0) lib/active_support/notifications/instrumenter.rb:20:in `instrument'
  activesupport (4.2.0) lib/active_support/notifications.rb:164:in `instrument'
  actionpack (4.2.0) lib/action_controller/metal/instrumentation.rb:30:in `process_action'
  actionpack (4.2.0) lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
  activerecord (4.2.0) lib/active_record/railties/controller_runtime.rb:18:in `process_action'
  actionpack (4.2.0) lib/abstract_controller/base.rb:137:in `process'
  actionview (4.2.0) lib/action_view/rendering.rb:30:in `process'
  actionpack (4.2.0) lib/action_controller/metal.rb:195:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
  actionpack (4.2.0) lib/action_controller/metal.rb:236:in `block in action'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `call'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:73:in `dispatch'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:42:in `serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:43:in `block in serve'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `each'
  actionpack (4.2.0) lib/action_dispatch/journey/router.rb:30:in `serve'
  actionpack (4.2.0) lib/action_dispatch/routing/route_set.rb:802:in `call'
  warden (1.2.3) lib/warden/manager.rb:35:in `block in call'
  warden (1.2.3) lib/warden/manager.rb:34:in `catch'
  warden (1.2.3) lib/warden/manager.rb:34:in `call'
  rack (1.6.1) lib/rack/etag.rb:24:in `call'
  rack (1.6.1) lib/rack/conditionalget.rb:38:in `call'
  rack (1.6.1) lib/rack/head.rb:13:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/flash.rb:260:in `call'
  rack (1.6.1) lib/rack/session/abstract/id.rb:225:in `context'
  rack (1.6.1) lib/rack/session/abstract/id.rb:220:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/cookies.rb:560:in `call'
  activerecord (4.2.0) lib/active_record/query_cache.rb:36:in `call'
  activerecord (4.2.0) lib/active_record/connection_adapters/abstract/connection_pool.rb:647:in `call'
  activerecord (4.2.0) lib/active_record/migration.rb:378:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `call'
  activesupport (4.2.0) lib/active_support/callbacks.rb:88:in `_run_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:734:in `_run_call_callbacks'
  activesupport (4.2.0) lib/active_support/callbacks.rb:81:in `run_callbacks'
  actionpack (4.2.0) lib/action_dispatch/middleware/callbacks.rb:27:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/reloader.rb:73:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
  web-console (2.1.2) lib/web_console/middleware.rb:37:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  railties (4.2.0) lib/rails/rack/logger.rb:38:in `call_app'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `block in call'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `block in tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:26:in `tagged'
  activesupport (4.2.0) lib/active_support/tagged_logging.rb:68:in `tagged'
  railties (4.2.0) lib/rails/rack/logger.rb:20:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/request_id.rb:21:in `call'
  rack (1.6.1) lib/rack/methodoverride.rb:22:in `call'
  rack (1.6.1) lib/rack/runtime.rb:18:in `call'
  activesupport (4.2.0) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  rack (1.6.1) lib/rack/lock.rb:17:in `call'
  actionpack (4.2.0) lib/action_dispatch/middleware/static.rb:113:in `call'
  rack (1.6.1) lib/rack/sendfile.rb:113:in `call'
  railties (4.2.0) lib/rails/engine.rb:518:in `call'
  railties (4.2.0) lib/rails/application.rb:164:in `call'
  rack (1.6.1) lib/rack/lock.rb:17:in `call'
  rack (1.6.1) lib/rack/content_length.rb:15:in `call'
  rack (1.6.1) lib/rack/handler/webrick.rb:89:in `service'
  /Users/holden/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/webrick/httpserver.rb:138:in `service'
  /Users/holden/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/webrick/httpserver.rb:94:in `run'
  /Users/holden/.rvm/rubies/ruby-2.2.0/lib/ruby/2.2.0/webrick/server.rb:294:in `block in start_thread'
```

**Comment by wingrunr21** [comment]
Stack trace?

**Comment by holden** [comment]
Added above.  Thanks,

**Comment by holden** [comment]
Something to do with encoding?

```
NoMethodError in Griddler::EmailsController#create
undefined method `valid_encoding?' for #<Array:0x007fd4aeda2a70>
Extracted source (around line #83):
81
82
83
84
85
86
              

    def clean_invalid_utf8_bytes(text)
      if text && !text.valid_encoding?
        text.force_encoding('ISO-8859-1').encode('UTF-8')
      else
        text
```

**Comment by holden** [comment]
I also tried uninstalling and reinstalling the gems.  v1.0.2 still works fine.  v1.1.0 breaks.

I can even leave the process method blank and it still breaks.

def process
   #why?????
end

**Comment by wingrunr21** [comment]
I'm looking into this but it is either a bug in `griddler` or the way we are handing something off to `griddler`.

**Comment by wouterw** [comment]
Looks like an issue in `griddler` itself. It doesn't play well with arrays as values in the `headers` hash.
It assumes all values are strings which have a [`valid_encoding?`](http://ruby-doc.org/core-2.1.0/String.html#method-i-valid_encoding-3F) method. Reworking [`extract_headers`](https://github.com/thoughtbot/griddler/blob/master/lib/griddler/email.rb#L51-L60) in the `Griddler::Email` class should do the trick.

**Comment by gabebw** [comment]
I just released Griddler version 1.2.1, which includes the code from https://github.com/thoughtbot/griddler/pull/212.

**Comment by wingrunr21** [comment]
@holden @wouterw does Griddler v1.2.1 solve the problem? If yes, I'll bump the min version in the gemspec and push a version bump.

**Comment by holden** [comment]
Yes it does. It works great now.

**Comment by wingrunr21** [comment]
Ok, I just released v1.1.1 and yanked v1.1.0 since it was not working. Thanks everyone!
